### PR TITLE
Allows roundstart observers to freely use antag HUD without losing respawnability

### PR DESCRIPTION
**Pull request by lewcc**
## What Does This PR Do
Title, allows roundstart observers to freely toggle AntagHUD, without losing respawnability.
### This only applies to people who press the Observe button on the new player panel before the game officially starts.
Also, improves admin logging around the usage of antag hud, particularly by adding a new two lines to the player panel indicating if a user has antag hudded, and if they were a roundstart observer. Also, if you antag hud and join as a pAI or a robobrain, admins will be notified.
Lastly, on the code side, this does away with checking for the `has_enabled_antaghud` variable tracked on ghosts, and moves it out to a global list instead.

## Why It's Good For The Game
If you join as a roundstart observer, I don't think it's much of a stretch to say that you already have de-facto antag HUD due to the amount of meta-knowledge that you get from other observers. People talk, and there are many ways to see what's up without even enabling the HUD.
Anyway, by clicking Observe before the round even starts, you've already committed to sitting out of the round without any knowledge of what you're joining onto at all (not even alert level), and this is basically your ready-up. It couldn't hurt to give these roundstart observers a bit more information on the game they've committed to watching, instead of playing.

In the end, this shouldn't really change much, it'll just make it more convenient for the few people who choose from the get-go to sit back and watch the game for a round.

## Testing
Loaded in, de-admined myself, and verified that I lost respawnability when I should, and that I didn't when I shouldn't.
Also checked the new player panel setup.

## Changelog
:cl:
tweak: If you observe before the round starts, you can toggle AntagHUD without losing respawnability.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/brain/robotic_brain.dm [hunk 1]
```diff
@@ -121,6 +121,9 @@
 	to_chat(brainmob, "<b>Remember, the purpose of your existence is to serve [imprinted_master]'s every word, unless lawed  or placed into a mech in the future.</b>")
 	brainmob.mind.assigned_role = "Positronic Brain"
 
+	if(brainmob.ckey in GLOB.antag_hud_users)
+		log_and_message_admins("[key_name(brainmob)] has joined as a robot brain, after having toggled antag hud.")
```
Comment (Contrabang) [inline code/modules/mob/living/carbon/brain/robotic_brain.dm:125]: log_and_message_admins already uses key_name(usr) in here, you'll probably want to refactor this to just be a log and message seperately

File: code/modules/mob/mob_misc_procs.dm [hunk 2]
```diff
@@ -139,9 +139,9 @@
 
 /proc/cannotPossess(A)
 	var/mob/dead/observer/G = A
-	if(G.has_enabled_antagHUD && GLOB.configuration.general.restrict_antag_hud_rejoin)
-		return 1
-	return 0
+	if(GLOB.configuration.general.restrict_antag_hud_rejoin && !(G.ckey in GLOB.roundstart_observer_keys) && (G.key in GLOB.antag_hud_users))
```
Comment (Contrabang) [inline code/modules/mob/mob_misc_procs.dm:142]: This is a long line, can you make part of this a guard clause

File: code/modules/mob/new_player/new_player.dm [hunk 3]
```diff
@@ -186,8 +186,9 @@
 			src << browse(null, "window=playersetup")
 			spawning = TRUE
 			stop_sound_channel(CHANNEL_LOBBYMUSIC)
-
-
+			if(SSticker.current_state < GAME_STATE_PLAYING)
```
Comment (Contrabang) [inline code/modules/mob/new_player/new_player.dm:189]: I'd say you should give this a 1-5 minute timer for ghosts to join imo. Maybe 2 minutes

File: code/modules/tgui/modules/ghost_hud_panel.dm [hunk 4]
```diff
@@ -62,14 +62,20 @@ GLOBAL_DATUM_INIT(ghost_hud_panel, /datum/ui_module/ghost_hud_panel, new)
...
 				var/response = alert(ghost, "If you turn this on, you will not be able to take any part in the round.", "Are you sure you want to enable antag HUD?", "Yes", "No")
 				if(response == "No")
 					return FALSE
 
-				ghost.has_enabled_antagHUD = TRUE
 				ghost.can_reenter_corpse = FALSE
 				REMOVE_TRAIT(ghost, TRAIT_RESPAWNABLE, GHOSTED)
+				log_and_message_admins("[key_name(ghost)] has enabled antaghud as an observer and forfeited respawnability.")
+
+			else if(ghost.ckey in GLOB.roundstart_observer_keys && !(ghost.ckey in GLOB.antag_hud_users))
```
Comment (Contrabang) [inline code/modules/tgui/modules/ghost_hud_panel.dm:74]: these checks are kind long no matter where they are. Could you make it into a proc on observers?

**Review by Contrabang — CHANGES_REQUESTED** [review]

File: code/modules/mob/new_player/new_player.dm (hunk 3 above)
Comment (lewcc) [reply code/modules/mob/new_player/new_player.dm:189]: Ehh, maybe in a later PR. For now, just roundstart is what's been approved.

File: code/modules/mob/dead/observer/observer_base.dm [hunk 5]
```diff
@@ -790,3 +790,22 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 
 	var/datum/spawners_menu/menu = new /datum/spawners_menu(src)
 	menu.ui_interact(src)
+
+/**
+ * Check if a player has antag-hudded, and if so, if they can rejoin.
+ * Returns TRUE if the player can rejoin, and FALSE if the player is ineligible to rejoin.
+ * If allow_roundstart_observers is FALSE (TRUE by default), then any observers who were able to ahud due to joining roundstart will be excluded as well.
+ */
+/mob/dead/observer/proc/check_ahud_rejoin_eligibility(allow_roundstart_observers=TRUE)
```
Comment (Contrabang) [inline code/modules/mob/dead/observer/observer_base.dm:799]:
```suggestion
/mob/dead/observer/proc/check_ahud_rejoin_eligibility(allow_roundstart_observers = TRUE)
```

File: code/modules/tgui/modules/ghost_hud_panel.dm [hunk 6]
```diff
@@ -62,14 +62,20 @@ GLOBAL_DATUM_INIT(ghost_hud_panel, /datum/ui_module/ghost_hud_panel, new)
...
-			if(!check_rights(R_ADMIN | R_MOD, FALSE) && !ghost.has_enabled_antagHUD && GLOB.configuration.general.restrict_antag_hud_rejoin)
+			if(!check_rights(R_ADMIN | R_MOD, FALSE) && !ghost.is_roundstart_observer() && GLOB.configuration.general.restrict_antag_hud_rejoin && !ghost.has_ahudded())
 				var/response = alert(ghost, "If you turn this on, you will not be able to take any part in the round.", "Are you sure you want to enable antag HUD?", "Yes", "No")
 				if(response == "No")
 					return FALSE
 
-				ghost.has_enabled_antagHUD = TRUE
 				ghost.can_reenter_corpse = FALSE
 				REMOVE_TRAIT(ghost, TRAIT_RESPAWNABLE, GHOSTED)
+				log_and_message_admins("[key_name(ghost)] has enabled antaghud as an observer and forfeited respawnability.")
```
Comment (Contrabang) [inline code/modules/tgui/modules/ghost_hud_panel.dm:72]: log_and_message_admins has an inherient `key_name(usr)` in it. You may want to split this up

**Review by Contrabang — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/silicon/pai/pai.dm [hunk 7]
```diff
@@ -110,6 +110,9 @@
 	var/datum/data/pda/app/messenger/M = pda.find_program(/datum/data/pda/app/messenger)
 	M.toff = TRUE
 
+	if(has_ahudded())
+		message_admins("[key_name(src)] has joined as a pAI, having previously enabled antag hud.")
```
Comment (Contrabang) [inline code/modules/mob/living/silicon/pai/pai.dm:114]: probably want a log_admin for this too?

**Review by Contrabang — CHANGES_REQUESTED** [review]

**Review by Contrabang — APPROVED** [review]
LGTM

**Review by S34NW — APPROVED** [review]

**Pull request merged**